src/provable/binable.unit-test.ts

```diff
@@ -1,32 +1,152 @@
-let uint64s = [...uint32s, 0x1000_0000_ffff_ffffn, 0xffff_ffff_ffff_ffffn];
+// uint
 
-for (let uint32 of uint32s) {
+test(Random.uint32, (uint32) => {
   let bytes = BinableUint32.toBytes(uint32);
   let result = BinableUint32.fromBytes(bytes);
   expect(result).toEqual(uint32);
-}
-for (let uint64 of uint64s) {
+});
+test(Random.uint64, (uint64) => {
   let bytes = BinableUint64.toBytes(uint64);
   let result = BinableUint64.fromBytes(bytes);
   expect(result).toEqual(uint64);
-}
-
-let noUint64s = [
-  -1n,
-  -10n,
-  -0xffff_ffff_ffff_ffffn,
-  0x1_0000_0000_0000_0000n,
-  0x1_2345_789a_bcde_ffffn,
-];
-let noUint32s = [...noUint64s, 0x1_0000_0000n, 0xffff_ffff_ffffn];
-
-for (let noUint32 of noUint32s) {
-  expect(() => BinableUint32.toBytes(noUint32)).toThrow('uint32 out of range');
-}
-for (let noUint64 of noUint64s) {
-  expect(() => BinableUint64.toBytes(noUint64)).toThrow('uint64 out of range');
-}
+});
+
+test.negative(Random.uint32.invalid, BinableUint32.toBytes);
+test.negative(Random.uint64.invalid, BinableUint64.toBytes);
+
+// finite fields
+
+test(Random.field, (field) => {
+  let bytes = Field.toBytes(field);
+  let result = Field.fromBytes(bytes);
+  expect(result).toEqual(field);
+});
+test.negative(Random.field.invalid, (field) =>
+  // TODO: toBytes doesn't fail. But should it?
+  Field.fromBytes(Field.toBytes(field))
+);
+
+test(Random.scalar, (scalar) => {
+  let bytes = Scalar.toBytes(scalar);
+  let result = Scalar.fromBytes(bytes);
+  expect(result).toEqual(scalar);
+});
+test.negative(Random.scalar.invalid, (scalar) =>
+  Scalar.fromBytes(Scalar.toBytes(scalar))
+);
+
+// public key
+
+test(Random.publicKey, (publicKey) => {
+  let bytes = PublicKey.toBytes(publicKey);
+  let result = PublicKey.fromBytes(bytes);
+  expect(result).toEqual(publicKey);
+});
+test.negative(Random.publicKey.invalid, (publicKey) =>
+  PublicKey.fromBytes(PublicKey.toBytes(publicKey))
+);
+
+// string
+
+test(Random.string(Random.nat(100)), (string) => {
+  let bytes = BinableString.toBytes(string);
+  let result = BinableString.fromBytes(bytes);
+  expect(result).toEqual(string);
+});
+
+// withVersionNumber
+
+let VersionedFieldV1 = withVersionNumber(Field, 1);
+let VersionedFieldV2 = withVersionNumber(Field, 2);
+
+test(Random.field, (field) => {
+  let bytes = VersionedFieldV1.toBytes(field);
+  let result = VersionedFieldV1.fromBytes(bytes);
+  expect(result).toEqual(field);
+});
+test.negative(Random.field, (field) => {
+  let bytes = VersionedFieldV2.toBytes(field);
+  // decoding with wrong version number should fail
+  let result = VersionedFieldV1.fromBytes(bytes);
+  expect(result).toEqual(field);
+});
+
+// check
+
+let FieldWithFailingCheck = withCheck(Field, () => {
+  throw Error('always fails');
+});
+test.negative(Random.field, (field) =>
+  // should fail to decode when check fails
+  FieldWithFailingCheck.fromBytes(FieldWithFailingCheck.toBytes(field))
+);
+
+let notABool = Random.map(Random.uint32, (x) => x + 2n);
```

Comment:
what if randomly `x = uint32::max - 1`

Reply:
the + is not modulo something, so `x + 2n` would then just exceed the max uint32, which is fine for this test